### PR TITLE
Set multiprocess_mode for Prometheus gauge metrics 

### PR DESCRIPTION
**Pull request by sravfeyn**
https://dimagi-dev.atlassian.net/browse/IIO-855

@snopoke

### CODE REVIEW

**Comment by dannyroberts** [comment]
This param is just ignored by the datadog backend, right?

I'm curious how you decided which metrics to use `multiprocess_mode='all'` versus `multiprocess_mode='max'` on. If `multiprocess_mode='max'` is most commonly what you want, maybe it's worth change that to be the default arg value instead, though maybe that's not the case.

**Comment by sravfeyn** [comment]
> This param is just ignored by the datadog backend, right?

Yes, because Datadog backend has this `accepted_gauge_params` [set to empty list](https://github.com/dimagi/commcare-hq/pull/27937/files#diff-5f012735a677866302f4169e2bf9d553R45)

> I'm curious how you decided which metrics to use multiprocess_mode='all' versus multiprocess_mode='max' on. If multiprocess_mode='max' is most commonly what you want, maybe it's worth change that to be the default arg value instead, though maybe that's not the case.

It's hard to decide the right param for various metrics. The aggregation is done over the same metric reported by multiple processes. So, if it's metric like `number_of_users_registered`, I would sum it, while if it's server_up duration then I would max it. We will probably not get these right and are planning to tweak them with trial and error. 

Re. default param, `all` is default one used by Prometheus if a value is not set. And using that breaks the metrics inside Prometheus which is an indication for us to go and pick a right aggregation.. We could still make max the default though

File: corehq/util/metrics/prometheus.py [hunk 1]
```diff
@@ -21,17 +21,30 @@ class PrometheusMetrics(HqMetrics):
...
     def _counter(self, name: str, value: float = 1, tags: Dict[str, str] = None, documentation: str = ''):
         """See https://prometheus.io/docs/concepts/metric_types/#counter"""
         try:
             self._get_metric(PCounter, name, tags, documentation).inc(value)
         except ValueError:
             pass
 
-    def _gauge(self, name: str, value: float, tags: Dict[str, str] = None, documentation: str = ''):
-        """See https://prometheus.io/docs/concepts/metric_types/#histogram"""
+    def _gauge(self, name: str, value: float, tags: Dict[str, str] = None, documentation: str = '', multiprocess_mode='all'):
```
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/util/metrics/prometheus.py:35]: Ditto previous nit. Adding the same suggestion here for convenience.
```suggestion
    def _gauge(
        self, 
        name: str, 
        value: float, 
        tags: Optional[Dict[str, str]] = None, 
        documentation: str = '', 
        multiprocess_mode: str = 'all',
    ):
```

File: corehq/util/metrics/prometheus.py [hunk 2]
```diff
@@ -21,17 +21,30 @@ class PrometheusMetrics(HqMetrics):
...
+        """
+        See https://prometheus.io/docs/concepts/metric_types/#histogram
+
+        multiprocess_mode: can be one of below values
```
Comment (kaapstorm) [inline corehq/util/metrics/prometheus.py:39]: Could be useful to use constants for these.

**Review by kaapstorm — APPROVED** [review]
Nothing blocking.

File: corehq/util/metrics/prometheus.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/util/metrics/prometheus.py:35]: On second thoughts ... I would drop the type annotations here, and move the docstring to the public method. The public shouldn't be using this method, so the docstring and the type annotations shouldn't help them here.

File: corehq/util/metrics/__init__.py [hunk 3]
```diff
@@ -165,9 +165,14 @@ def metrics_counter(name: str, value: float = 1, tags: Dict[str, str] = None, do
     provider.counter(name, value, tags=tags, documentation=documentation)
 
 
-def metrics_gauge(name: str, value: float, tags: Dict[str, str] = None, documentation: str = ''):
+def metrics_gauge(name: str, value: float, tags: Dict[str, str] = None, documentation: str = '', multiprocess_mode='all'):
```
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/util/metrics/__init__.py:168]: lol, a nit for my nit: Considering `multiprocess_mode` has a set of valid values, there's a type annotation for that ...
```suggestion
def metrics_gauge(
    name: str, 
    value: float, 
    tags: Dict[str, str] = None, 
    documentation: str = '', 
    multiprocess_mode: Literal['all', 'liveall', 'livesum', 'max', 'min'] = 'all',
):
```
Now your fancier IDEs will highlight invalid values. :smirk:

File: corehq/util/metrics/prometheus.py (hunk 2 above)
Comment (snopoke) [reply corehq/util/metrics/prometheus.py:39]: +1

**Review by snopoke — APPROVED** [review]
Nothing to add to other reviewers comments

**Review by snopoke — APPROVED** [review]

File: corehq/util/metrics/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/util/metrics/__init__.py:168]: Thanks Norman, will update.

File: corehq/util/metrics/prometheus.py (hunk 2 above)
Comment (sravfeyn) [reply corehq/util/metrics/prometheus.py:43]: Updated.

File: corehq/util/metrics/prometheus.py [hunk 4]
```diff
@@ -21,17 +24,31 @@ class PrometheusMetrics(HqMetrics):
...
         """See https://prometheus.io/docs/concepts/metric_types/#counter"""
         try:
             self._get_metric(PCounter, name, tags, documentation).inc(value)
         except ValueError:
             pass
 
-    def _gauge(self, name: str, value: float, tags: Dict[str, str] = None, documentation: str = ''):
-        """See https://prometheus.io/docs/concepts/metric_types/#histogram"""
+    def _gauge(self, name: str, value: float, tags: Dict[str, str]=None, documentation: str = '',
+               multiprocess_mode: Literal[MPM_ALL, MPM_LIVEALL, MPM_LIVESUM, MPM_MAX, MPM_MIN]=MPM_ALL):
```
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/util/metrics/prometheus.py:39]: :sob: Oh no. `Literal` only arrives in Python 3.8. Urgh, sorry @sravfeyn !

[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/util/metrics/prometheus.py]: Np, removed it

**Pull request merged**